src/librustc_lint/builtin.rs

```diff
@@ -2153,123 +2153,208 @@ impl ClashingExternDeclarations {
-                            Self::structurally_same_type(cx, a, b, ckind)
-                        })
-                        && Self::structurally_same_type(cx, a_sig.output(), b_sig.output(), ckind)
-                }
-                (Tuple(a_substs), Tuple(b_substs)) => {
-                    a_substs.types().eq_by(b_substs.types(), |a_ty, b_ty| {
-                        Self::structurally_same_type(cx, a_ty, b_ty, ckind)
-                    })
-                }
-                // For these, it's not quite as easy to define structural-sameness quite so easily.
-                // For the purposes of this lint, take the conservative approach and mark them as
-                // not structurally same.
-                (Dynamic(..), Dynamic(..))
-                | (Error(..), Error(..))
-                | (Closure(..), Closure(..))
-                | (Generator(..), Generator(..))
-                | (GeneratorWitness(..), GeneratorWitness(..))
-                | (Projection(..), Projection(..))
-                | (Opaque(..), Opaque(..)) => false,
-
-                // These definitely should have been caught above.
-                (Bool, Bool) | (Char, Char) | (Never, Never) | (Str, Str) => unreachable!(),
-
-                // An Adt and a primitive type. This can be FFI-safe is the ADT is an enum with a
-                // non-null field.
-                (Adt(..), other_kind) | (other_kind, Adt(..))
-                    if is_primitive_or_pointer(other_kind) =>
-                {
-                    let (primitive, adt) =
-                        if is_primitive_or_pointer(&a.kind) { (a, b) } else { (b, a) };
-                    if let Some(ty) = crate::types::repr_nullable_ptr(cx, adt, ckind) {
-                        ty == primitive
-                    } else {
-                        compare_layouts(a, b)
-                    }
+                                } else {
+                                    unreachable!()
+                                }
+                            }
+                            (Array(a_ty, a_const), Array(b_ty, b_const)) => {
+                                // For arrays, we also check the constness of the type.
+                                a_const.val == b_const.val
+                                    && structurally_same_type_impl(
+                                        seen_types, cx, a_const.ty, b_const.ty, ckind,
+                                    )
+                                    && structurally_same_type_impl(
+                                        seen_types, cx, a_ty, b_ty, ckind,
+                                    )
+                            }
+                            (Slice(a_ty), Slice(b_ty)) => {
+                                structurally_same_type_impl(seen_types, cx, a_ty, b_ty, ckind)
+                            }
+                            (RawPtr(a_tymut), RawPtr(b_tymut)) => {
+                                a_tymut.mutbl == b_tymut.mutbl
```

Comment:
while not part of this PR I would expect `*const T` and `*mut T` to be structurally equal.

To my knowledge they only differ in variance

Reply:
I see. My reasoning for having them not equal is because I want this case to clash:

```rust
mod a {
    extern "C" {
        fn take(_: *const S);
    }
}
mod b {
    extern "C" {
        fn take(_: *mut S);
    }
}
```
Specifically, I want to guard against the case where `take` actually mutates the data underneath, but  `a::take` is called with data that was supposed to be immutable -- a conflicting declaration of the mutability of some data should ring some alarm bells.